### PR TITLE
Sound rework

### LINKED ISSUE #505: wake button doesn't work when the music is playing or when the device is in busy state
#### Describe the bug
- Wake button(on respeaker hat) doesn't work when the device is in the busy state. wake button can be required for example to stop/pause a song. (hotword sensitivity is also lowered in the busy state).

#### To Reproduce
use wake button while playing a song

#### Expected behaviour
wake button should work in the busy state

#### Desktop (please complete the following information):
 - OS: susibian

### PR DESCRIPTION
**Pull request by alok760**
Fixes #505 #496

#### Checklist

- [x] I have read the Contribution & Best practices Guide and my PR follows them.
- [x] My branch is up-to-date with the Upstream master branch.
- [ ] The unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Test Passing

- [x] The SUSI Server must be building on the pi on bootup
- [x] The hotword detection should have a decent accuracy
- [ ] SUSI Linux shouldn't crash when switching from online to offline and vice versa (failing as of now)
- [ ] SUSI Linux should be able to boot offline when no internet connection available (failing)

#### Short description of what this resolves:
remove the secondary query processing stage and implement things in the IDLE state completely
- Music successfully stops on saying "susi, stop" and the player doesn't hang afterwards
- Low sensitivity for hotword detection issue is gone which was due to the secondary query stage
- Wake button now works while music is being played and can be used to ask queries
- Music playback stops when a query is being asked and resumes afterwards from the IDLE state

### CODE REVIEW

File: main/states/idle_state.py [hunk 1]
```diff
@@ -43,11 +44,23 @@ def on_enter(self, payload=None):
         self.notify_renderer('idle')
 
     def __detected(self):
-        if self.isActive:
+        if hasattr(self, 'video_process') and self.video_process != None:
```
Comment (norbusan) [inline main/states/idle_state.py]: You only check for a video process, but not for an audio process, please update this, too

Comment (alok760) [reply main/states/idle_state.py]: solved in the latest commit

File: main/states/base_state.py [hunk 2]
```diff
@@ -11,7 +11,8 @@ class State(ABC):
     """State is an abstract class to represent a state. It defines the abstract method for behavior of state and
     transition method to facilitate transition between states.
     """
-
+    video_process = None
```
Comment (stealthanthrax) [inline main/states/base_state.py]: Follow a getter/setter approach here and use the `@property` i.e. declare these in init and do not access these directly by the class name.

File: main/states/busy_state.py [hunk 3]
```diff
@@ -107,22 +80,26 @@ def on_enter(self, payload=None):
...
+                    #stopAction.run()
+                    #stopAction.detector.terminate()
 
                 else:
                     audio_url = reply['identifier']
                     audio_process = subprocess.Popen(['play', audio_url[6:], '--no-show-progress'])  # nosec #pylint-disable type: ignore
                     self.audio_process = audio_process
-                    stopAction.run()
-                    stopAction.detector.terminate()
+                    State.audio_process = audio_process
```
Comment (stealthanthrax) [inline main/states/busy_state.py]: here as well

File: main/states/busy_state.py [hunk 4]
```diff
@@ -107,22 +80,26 @@ def on_enter(self, payload=None):
                         url = data['url']
                         video_process = subprocess.Popen(['cvlc', 'https' + url[5:], '--no-video'])
                         self.video_process = video_process
+                        State.video_process = video_process
```
Comment (stealthanthrax) [inline main/states/busy_state.py]: change here as well

**Comment by stealthanthrax** [comment]
@alok760 , please look at the comments. Rest looks good.

File: main/states/idle_state.py [hunk 5]
```diff
@@ -43,11 +44,33 @@ def on_enter(self, payload=None):
...
+        elif hasattr(self, 'audio_process') and self.audio_process != None:
+            self.audio_process.send_signal(signal.SIGSTOP)  # nosec #pylint-disable type: ignore
+            lights.off()
+            lights.wakeup()
+            subprocess.Popen(['play', os.path.join(self.components.config['data_base_dir'],
+                                                   self.components.config['detection_bell_sound'])])  # nosec #pylint-disable type: ignore
+            lights.wakeup()
+            self.transition(self.allowedStateTransitions.get('recognizing'))
+            self.audio_process.send_signal(signal.SIGCONT)  # nosec #pylint-disable type: ignore
+
```
Comment (norbusan) [inline main/states/idle_state.py]: Can we get rid of this duplication of code? Besides the process that is the signal sent to it is the same code, so it shouldn't be duplicated.

Comment (alok760) [reply main/states/idle_state.py]: updated

**Comment by stealthanthrax** [comment]
@alok760 , why is the `video_process` code removed?

**Comment by alok760** [comment]
> @alok760 , why is the `video_process` code removed?

@stealthanthrax 
i used the same process name that enables -
- only one sound process at a time which prevents running of sound from youtube stream and local audio stream at the same time 
- removal of duplicate code

**Pull request merged**